### PR TITLE
Ext pillar config param gh

### PR DESCRIPTION
**Pull request by alexbleotu**
### What does this PR do?

1. Added the ability to pass extra minion data in a new optional parameter in external pillar functions called `extra_minion_data`; this is fully backwards compatible - the external pillars functions are inspected before they are called and the new parameter is only passed if explicitly defined. 
1. Added the `pass_to_ext_pillars` key in the minion config that allows to pass minion config data in the `extra_minion_data` parameter to enrich the information about the minion the external pillar is invoked for. This allows to better control the data retrieved by the external pillars and is very useful if the ``minion_ids`` have a flat naming scheme.
 
### What issues does this PR fix or reference?
saltstack/morganstanley#20

### Previous Behavior
The only minion specific info external pillar functions was the ``minion_id``

### New Behavior
The new `extra_minion_data` parameter can be defined to pass more data to external pillar functions. The `pass_to_ext_pillar` config key can be used to specify which config keys are to be passed to external pillar functions in the `extra_minion_data` parameter.

### Tests written?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by alexbleotu** [comment]
@cachedout  @thatch45 This is the replacement for https://github.com/saltstack/salt/pull/42738. Hope this is ok. I will fix the conflicts asap.

**Comment by alexbleotu** [comment]
Fixed conflicts, pylint, tests

**Comment by alexbleotu** [comment]
Hey everybody. Any chance we can get this merged in?

**Comment by rallytime** [comment]
@cachedout and @terminalmage bump :)

**Review by cachedout — APPROVED** [review]
I am OK with this but I would like @terminalmage and @thatch45 to have a say.

**Review by terminalmage — APPROVED** [review]

File: salt/pillar/__init__.py [hunk 1]
```diff
@@ -768,17 +841,35 @@ def _external_pillar_data(self, pillar, val, key):
         Builds actual pillar data structure and updates the ``pillar`` variable
         '''
         ext = None
+        args = inspect.getargspec(self.ext_pillars[key]).args
```
Comment (thatch45) [inline salt/pillar/__init__.py]: @cachedout does six or some other code handle the fact that `inspect.getargspec` is deprecated in python 3?

Comment (cachedout) [reply salt/pillar/__init__.py]: Guh. Good call. I totally forgot about that. IIRC, we've had to gate that behind a PY3-specific check and branch into the proper call.

**Review by thatch45 — APPROVED** [review]

File: salt/pillar/__init__.py (hunk 1 above)
Comment (cachedout) [reply salt/pillar/__init__.py]: @alexbleotu We need to check if we are running Py3 and if so, switch to [inspect.getfullargspec()](https://docs.python.org/3.6/library/inspect.html#inspect.getfullargspec)

**Review by cachedout — CHANGES_REQUESTED** [review]
Needs one change for deprecated function in python3

File: salt/pillar/__init__.py (hunk 1 above)
Comment (terminalmage) [reply salt/pillar/__init__.py]: We have a convenience function for this: `salt.utils.args.get_function_argspec()`

Comment (alexbleotu) [reply salt/pillar/__init__.py]: Excellend. Will do

Comment (alexbleotu) [reply salt/pillar/__init__.py]: Done

**Comment by alexbleotu** [comment]
@cachedout Done the changes

**Review by cachedout — APPROVED** [review]

**Pull request merged**